src/components/astra-editor/index.ts

```diff
@@ -0,0 +1,351 @@
+      }
+
+      .cm-completionIcon-function::after,
+      .cm-completionIcon-method::after,
+      .cm-completionIcon-variable::after,
+      .cm-completionIcon-namespace::after,
+      .cm-completionIcon-interface::after {
+        content: '⚡' !important;
+      }
+    `
+
+    this.shadowRoot?.append(styleElement)
+
+    this.container = container
+  }
+
+  connectedCallback(): void {
+    this._color = this.getAttribute('color') === 'dark' ? 'dark' : 'light'
+    this._theme = this.getAttribute('theme') ?? 'moondust'
+    this._placeholder = this.getAttribute('placeholder') ?? ''
+
+    if (this._color === 'dark') {
+      this.container.classList.add('dark')
+    } else {
+      this.container.classList.remove('dark')
+    }
+
+    // Default extensions
+    this.extensions = [
+      {
+        name: 'basic-setup',
+        ext: [
+          lineNumbers(),
+          highlightActiveLineGutter(),
+          history(),
+          dropCursor(),
+          indentOnInput(),
+          syntaxHighlighting(defaultHighlightStyle, { fallback: true }),
+          autocompletion(),
+          highlightActiveLine(),
+          indentUnit.of('  '),
+          keymap.of([
+            {
+              key: 'Tab',
+              preventDefault: true,
+              run: (target) => {
+                if (completionStatus(target.state) === 'active') {
+                  acceptCompletion(target)
+                } else {
+                  indentMore(target)
+                }
+                return true
+              },
+              shift: indentLess,
+            },
+            {
+              key: 'Shift-Enter',
+              preventDefault: true,
+              run: insertNewline,
+            },
+            {
+              key: 'Ctrl-Space',
+              mac: 'Cmd-i',
+              preventDefault: true,
+              run: startCompletion,
+            },
+            ...closeBracketsKeymap,
+            ...defaultKeymap,
+            ...historyKeymap,
+            ...foldKeymap,
+          ]),
+        ],
+        comp: new Compartment(),
+      },
+      { name: 'placeholder', ext: placeholder(this._placeholder), comp: new Compartment() },
+      {
+        name: 'theme',
+        ext: getPredefineTheme(this._color, this._theme),
+        comp: new Compartment(),
+      },
+      {
+        name: 'onchange',
+        ext: EditorView.updateListener.of((vu: ViewUpdate) => {
+          if (vu.docChanged) {
+            const value = vu.state.doc.toString()
+            this._value = value
+            this.dispatchEvent(new CustomEvent('change', { detail: value }))
+          }
+        }),
+        comp: new Compartment(),
+      },
+    ]
+
+    const editor = new EditorView({
+      doc: this.getAttribute('value') ?? '',
+      extensions: this.getExtensions(),
+      parent: this.container,
+      root: this.shadowRoot as ShadowRoot,
+    })
+
+    this.editor = editor
+  }
+
+  render() {
+    return this.container
```

Comment:
i would expect to see the following here instead of manually creating a `<div />` tag above

```
return html`<div style="..." ... \>`
```

Reply:
I haven't touched LitElement in detail, but I need to grab the <div> reference in the connectedCallback. I am not sure which one execute first. `connectedCallback` or `render`

Reply:
they have `ref` similar to react you can use instead of trying to query the shadowRoot: 
https://lit.dev/docs/templates/directives/#ref

```js
@customElement('my-element')
class MyElement extends LitElement {

  inputRef: Ref<HTMLInputElement> = createRef();

  render() {
    // Passing ref directive a Ref object that will hold the element in .value
    return html`<input ${ref(this.inputRef)}>`;
  }

  firstUpdated() {
    const input = this.inputRef.value!;
    input.focus();
  }
}
```

Reply:
Got it. Let me try